File: src/main/java/net/asodev/islandutils/state/crafting/CraftingToast.java
package net.asodev.islandutils.state.crafting;

import net.asodev.islandutils.state.crafting.state.CraftingItem;
import net.minecraft.ChatFormatting;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.toasts.Toast;
import net.minecraft.client.gui.components.toasts.ToastComponent;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.Style;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;

import static net.asodev.islandutils.util.Utils.MCC_HUD_FONT;

public class CraftingToast implements Toast {
    private static ResourceLocation ISLAND_TOASTS_TEXTURE = new ResourceLocation("island", "textures/gui/toasts.png");

    ItemStack itemStack;
    Component displayName;
    final Component description = Component.literal("CRAFTING COMPLETE!")
            .withStyle(MCC_HUD_FONT.withColor(ChatFormatting.WHITE));

    public CraftingToast(CraftingItem craftingItem) {
        itemStack = craftingItem.getStack();
        displayName = craftingItem.getTitle();
    }

    @Override
    public Toast.Visibility render(GuiGraphics guiGraphics, ToastComponent toastComponent, long l) {
        guiGraphics.blit(ISLAND_TOASTS_TEXTURE, 0, 0, 0, 0, this.width(), this.height());
        int y = 7;
        guiGraphics.drawString(toastComponent.getMinecraft().font, description, 30, y, -16777216, false);
        y += 5 + 4;
        guiGraphics.drawString(toastComponent.getMinecraft().font, displayName, 30, y, -11534256, false);

        guiGraphics.renderFakeItem(itemStack, 8, 8);
        return (double)l >= 5000.0 * toastComponent.getNotificationDisplayTimeMultiplier() ? Toast.Visibility.HIDE : Toast.Visibility.SHOW;
    }
}

File: src/main/kotlin/dev/asodesu/islandutils/features/crafting/notif/CraftingToast.kt
package dev.asodesu.islandutils.features.crafting.notif

import dev.asodesu.islandutils.api.extentions.Resources
import dev.asodesu.islandutils.api.extentions.pose
import dev.asodesu.islandutils.features.crafting.items.CraftingItem
import net.minecraft.ChatFormatting
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.components.toasts.Toast
import net.minecraft.client.gui.components.toasts.ToastManager
import net.minecraft.client.renderer.RenderType
import net.minecraft.network.chat.Component
import net.minecraft.util.ARGB
import dev.asodesu.islandutils.Font as MccFont

class CraftingToast(private val craftingItem: CraftingItem) : Toast {
    private val BACKGROUND_SPRITE = Resources.islandUtils("toast/crafting_toast")
    private val VISIBLE_FOR = 5000L
    private val TITLE_TEXT = Component.translatable("islandutils.feature.crafting.complete")
        .withStyle(MccFont.HUD_STYLE.withColor(ChatFormatting.WHITE))

    private val TOP_PADDING = 7
    private val HUD_FONT_HEIGHT = 5
    private val TEXT_GAP = 4
    private val ITEM_HEIGHT = 16

    private var visibility: Toast.Visibility = Toast.Visibility.HIDE

    override fun getWantedVisibility() = visibility
    override fun update(toastManager: ToastManager, l: Long) {
        val visibleFor = VISIBLE_FOR * toastManager.notificationDisplayTimeMultiplier
        visibility = if (l > visibleFor) Toast.Visibility.HIDE else Toast.Visibility.SHOW
    }

    override fun render(guiGraphics: GuiGraphics, font: Font, l: Long) {
        guiGraphics.blitSprite(RenderType::guiTextured, BACKGROUND_SPRITE, 0, 0, this.width(), this.height())

        var y = TOP_PADDING
        guiGraphics.drawString(font, TITLE_TEXT, 30, y, ARGB.white(1f), false)
        y += HUD_FONT_HEIGHT + TEXT_GAP
        guiGraphics.drawString(font, craftingItem.item.hoverName, 30, y, ARGB.white(1f), false)

        guiGraphics.renderFakeItem(craftingItem.item, 8, 8)

        guiGraphics.pose {
            translate(0.0, 0.0, 200.0)
            guiGraphics.drawString(font, craftingItem.type.icon, 6, 8 + ITEM_HEIGHT - 6, ARGB.white(1f), false)
        }
    }
}